Update free domain suggestions UI

ovitrif opened the pull request:
Resolves #17899 

This PR:
- ⊕ Adds the new UI for domain suggestions using Compose #17899
- ⊕ Refactors view logic for domain suggestions:
  - ⊕ Migrates the domain suggestion view-holder to view binding
  - ⊕ Refactors the UI state of domain list items to separate the old vs new one
  - ⊖ Removes the repetitive `Domain` prefixes from ui state class names when it can be deduced
  - ⊜ Converts single line vm methods from block to expression body

> **Note**
> To easily review the many changes please go commit by commit.

## To Test
#### Prerequisites:

<details>
<summary>◐ Toggle the Site Creation Domain Purchasing feature (<code>SiteCreationDomainPurchasingFeatureConfig</code>)</summary>

1. Go to `Me` → `App Settings` → `Debug settings`
3. Scroll to the `Features in development` section
4. Tap on `SiteCreationDomainPurchasingFeatureConfig`
6. Tap `RESTART THE APP` button
---
</details>

### 🅰️ Control Variation

1. Open the Jetpack app and login
2. ◐ Toggle the `Site Creation Domain Purchasing` feature flag `OFF` (should be the case)
3. Start the Site Creation flow: `Open Site Picker` → `➕` → `Create WordPress.com site`
4. Continue the flow till `Choose a domain` screen
5. **Verify** the expected **results** functionality is maintained by the refactoring:
   1. Enter `test` in the search field
   2. **Expect** the first result to be the unavailable `test.wordpress.com` followed by available one(s)
   3. Tap on the `✗` button to clear the search query
   4. **Expect** The results to be replaced by the placeholder illustration
   5. Enter `tested` in the search field
   6. **Expect** all results to be available and selectable
6. **Verify** the expected **error** functionality keeps working:</summary>
   1. Tap on the `✗` button to clear the search query
   2. Turn off internet connectivity on your testing device
   3. Enter `testet` in the search field
   4. **Expect** The result to be `There is no network available`
   5. Turn on internet connectivity
   6. Tap the error item (e.g. on `Retry`)
   7. **Expect** the results to be updated
   8. Replace the `t` in the search field with `%`
   9. **Expect** the error message for invalid characters

### 🅱️ Treatment Variation

1. ◐ Toggle the `Site Creation Domain Purchasing` feature flag `ON`
7. Start the Site Creation flow: `Open Site Picker` → `➕` → `Create WordPress.com site`
8. Continue the flow till `Choose a domain` screen
9. **Verify** free and paid domains are suggested
10. Tap on a free domain
11. **Verify** the `Create Site` button is **displayed**
12. Tap on a paid domain
13. **Verify** the `Create Site` button is **hidden**

### Screenshots

| 🅰️ Control | 🅱️ Treatment |
| - | - |
| Before | After |
| <img width="300" src="https://user-images.githubusercontent.com/4588074/220423925-9ff34be3-3680-40e2-825b-3a94a442c4b6.png"> | <img width="300" src="https://user-images.githubusercontent.com/4588074/220423913-7ab5cc97-0215-4ab7-b96b-a8f573a0a657.png"> |

## Regression Notes
1. Potential unintended areas of impact
   Site creation → Domains when the experimental feature is disabled

3. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing

4. What automated tests I added (or what prevented me from doing so)
   Updated unit tests in `SiteCreationDomainsViewModelTest`. I don't think more tests are needed since most of the changes are in the view layer and the improved ui state strong typing is a solid way to guard against regressions when it comes to view-model.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Refactor: Migrate domain view-holder to view binding

ovitrif committed:
Refactor: Replace domain unavailable check with list.none

ovitrif committed:
Refactor: Move empty list message string logic inside ui string res

ovitrif committed:
Refactor: Remove braces from isStarted check

ovitrif committed:
Refactor: Convert simple methods to single-line expression body

ovitrif committed:
Refactor: Embed domains list item view type in ui state

ovitrif committed:
Refactor: Move onCreateViewHolder under onBindViewHolder followed by getItemViewType

ovitrif committed:
Refactor: Replace ui state's clickable flag w/ scoped onItemTapped fn

ovitrif committed:
Refactor: Rename ui state's onItemTapped to onClock

ovitrif committed:
Refactor: Add onClick to constructor of available domain ui state

ovitrif committed:
Refactor: Move all logic in domain suggestion item view-holder to onBind

ovitrif committed:
Refactor: Move all logic in domain suggestion error viewHolder to onBind

ovitrif committed:
Refactor: Remove redundant requireNotNull from domain error viewHolder

ovitrif committed:
Refactor: Pass specific uiState type to domain view holders

ovitrif committed:
Refactor: Cleanup and tidy logic in domain view holders

ovitrif committed:
WIP: Add compose view for new domains ui

ovitrif committed:
Refactor: Remove deducible qualifying prefixes from ui-state models

ovitrif committed:
Refactor: Separate old item ui state from the new one by sealed class

ovitrif committed:
Refactor: Remove temporary cost added to domain name

ovitrif committed:
Add and display raw domain cost

ovitrif committed:
Implement onClick for free domains

ovitrif committed:
Hide old unavailability item in new list ui

ovitrif committed:
Fix checkstyle errors

ovitrif committed:
Suppress ReturnCount check to fix detekt issues

ovitrif committed:
Add divider and gray cost text color

ovitrif committed:
Fix unit tests for old domains ui

ParaskP7 commented:
👋 @ovitrif !

I am seeing being included in this feature related PR review, alongside other feature engineers, was that on purpose, is there is anything core specific that you would like me to review here? 🤔

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainsAdapter.kt:
```diff
@@ -47,6 +53,15 @@ class SiteCreationDomainsAdapter(private val uiHelpers: UiHelpers) : Adapter<Sit
         diffResult.dispatchUpdatesTo(this)
     }
 
+    @Suppress("ForbiddenComment")
+    override fun onViewRecycled(holder: SiteCreationDomainViewHolder<*>) {
+        if (holder is DomainComposeItemViewHolder) {
+            // TODO: Remove this for Compose 1.2.0-beta02+ and RecyclerView 1.3.0-alpha02+
+            holder.composeView.disposeComposition()
+        }
+        super.onViewRecycled(holder)
+    }
+
```
@thomashorta I added you as reviewer to this PR but only to have someone experienced with `Compose` validate this commit: https://github.com/wordpress-mobile/WordPress-Android/pull/17998/commits/9a35659fec875d0b09ecab4795878e787391d4f2

@ParaskP7 I added you as reviewer to inform someone from `infrastructure (core)` of what the above commit entails (please read the code comments).

P.S. You don't have to really look at anything else from this PR, feel free to skip to the above commit and the only thing relevant for you there is in the comments.

**TL;DR:** I added Compose configuration which should be removed when we upgrade to `Compose 1.2.0-beta02+` and `RecyclerView 1.3.0-alpha02+`, because they provide better out-of-the-box support for Using Compose in a RecyclerView[^source].

[^source]: [Using Compose in a RecyclerView](https://medium.com/androiddevelopers/jetpack-compose-interop-using-compose-in-a-recyclerview-569c7ec7a583
)

ParaskP7 replied:
Oh great, thank you so much for doing so and for connecting the dots @ovitrif , super important! 🥇 

Btw, how about you referencing that on this #17563 issue too, I think it will help having that there, so that when I get unblock (by #17790), I won't forget about this detail and possible update `RecyclerView` to its latest stable version (or even `RC`, should it be needed).

1. Question (❓): Will updating to `Compose 1.2.0-beta02+` alone not be enough?
2. Question (❓): It seems that `RecyclerView` is already on [1.3.0-rc01](https://developer.android.com/jetpack/androidx/releases/recyclerview#recyclerview-1.3.0-rc01), but this release has been there for months now (`September 21, 2022`). I hope the production ready release for `1.3.0` gets out sooner than later? As such, and again, just to be explicit about it, will updating to `RecyclerView 1.3.0` alone not be enough?
3. Question (❓): Maybe we should include this `RecyclerView 1.3.0-alpha02+` update to become part of #17563 too, wdyt?

ovitrif replied:
@ParaskP7 Thanks a lot for following up on this with more feedback 🙇 !

I added [this comment](https://github.com/wordpress-mobile/WordPress-Android/issues/17563#issuecomment-1439830915) to issue #17563 👍 

To answer `1` and `2`, I'm not really sure if both dependencies versions should be higher but if we have to check when upgrading only one of them I'm willing to do [the "trick" explained](https://gist.github.com/arriolac/cf4966d6f02f5890f8ea0058831fcadf#file-itemrow-kt) in the [linked medium article](https://medium.com/androiddevelopers/jetpack-compose-interop-using-compose-in-a-recyclerview-569c7ec7a583) to count recompositions, which should answer this 👍 .

My answer to question `3` is to include this `RecyclerView` version 👍 . Should I add something for this to the issue?

ParaskP7 replied:
Great, thanks you for adding the comment there @ovitrif ! 🌟 

PS: I updated the description accordingly so that we don't forget this extra change when we update `Compose` at some point in the future.

> To answer 1 and 2, I'm not really sure if both dependencies versions should be higher but if we have to check when upgrading only one of them I'm willing to do [the "trick" explained](https://gist.github.com/arriolac/cf4966d6f02f5890f8ea0058831fcadf#file-itemrow-kt) in the [linked medium article](https://medium.com/androiddevelopers/jetpack-compose-interop-using-compose-in-a-recyclerview-569c7ec7a583) to count recompositions, which should answer this 👍 .

Awesome, this helps a lot and I'll revisit/test that when the time comes! 💯 

> My answer to question 3 is to include this RecyclerView version 👍 . Should I add something for this to the issue?

Coolio, noted, I think I agree, but no need to add something there, at least not yet. I'll revisit the `RecyclerView` update and the removal of `holder.composeView.disposeComposition()` upon completing the `Compose` updates. 🙏

ovitrif commented:
> 👋 @ovitrif !
> 
> I am seeing being included in this feature related PR review, alongside other feature engineers, was that on purpose, is there is anything core specific that you would like me to review here? 🤔

👋 Petros, you're uber fast 🚀 👀 🏆 
Please see [the comment I added afterwards](https://github.com/wordpress-mobile/WordPress-Android/pull/17998#discussion_r1114140142) 🙏 !

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainViewHolder.kt:
```diff
@@ -1,88 +1,81 @@
+            setContent {
+                AppTheme {
+                    DomainItem(uiState)
+                }
```
❤️ Nice! I was playing with adding a Compose item in the Dashboard `RecyclerView` exactly like you did it here a few weeks ago. But I ended up taking a step back since I was about to take 1 week off and didn't have the time to look into possible caveats. 

Thanks for looking into that and making the necessary changes for getting it to work properly!

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainViewHolder.kt:
```diff
@@ -1,88 +1,81 @@
+    }
+
+    @Suppress("ForbiddenComment")
+    class DomainComposeItemViewHolder(
```
🔍 Super small nit-picking: The class (`DomainComposeItemViewHolder`) could be renamed to express what it is (something that tells this is the **new** version or **v2**) instead of how it works (**Compose**). I think renaming the other class above (`DomainSuggestionItemViewHolder`) to mention **old** / **v1** might be worth it as well if you fell like this suggestion makes sense. Wdyt?

ovitrif replied:
This is actually a good point, I'm in favor of the `Old` vs `New` naming since it's already used and v1 looks ugly to me with `lowerCamelCaseV1` 😅

ovitrif replied:
👋 @thomashorta Addressed this in commit f63f3a70

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainsAdapter.kt:
```diff
@@ -4,54 +4,67 @@ import android.view.ViewGroup
+    override fun onViewRecycled(holder: SiteCreationDomainViewHolder<*>) {
+        if (holder is DomainComposeItemViewHolder) {
+            // TODO: Remove this for Compose 1.2.0-beta02+ and RecyclerView 1.3.0-alpha02+
+            holder.composeView.disposeComposition()
```
❤️ Again, thanks for making the necessary changes to make this work with the lib versions we have currently! I think this was the part that I was missing when I tried to use a `Composable` inside a `RecyclerView` a few weeks ago.

ovitrif replied:
Awesome! Glad to know it could be helpful not only or me 🙇

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainsViewModel.kt:
```diff
@@ -193,25 +181,21 @@ class SiteCreationDomainsViewModel @Inject constructor(
-                UiStringRes(R.string.new_site_creation_empty_domain_list_message)
-            }
+            val emptyListMessage = UiStringRes(
+                if (event.isError && event.error.type == SuggestDomainErrorType.INVALID_QUERY)
+                    R.string.new_site_creation_empty_domain_list_message_invalid_query
+                else R.string.new_site_creation_empty_domain_list_message
+            )
```
🔍 Nit-picking: I usually prefer using the curly braces in multi-line if-else statements, which adds a bit of clarity and is also recommended in the [Kotlin](https://kotlinlang.org/docs/coding-conventions.html#control-flow-statements) and [Android](https://developer.android.com/kotlin/style-guide#braces) style guides.

ovitrif replied:
Idk, the reason I'm formatting if-else's like this is because I actually miss the ternary operator (`tru ? dis : dat`) in Kotlin, probably the most significant thing it lacks 😅 . Tbh I prefer if else without curly braces if it's on one line (which is the case by my reasoning 😅 ), and this is (quite) widely used in the codebase.

- The [linked Kotlin style](https://kotlinlang.org/docs/coding-conventions.html#control-flow-statements) is concerned with **conditions** which are multiline 😉 .

- Why they recommend it in the [official Android style guide](https://developer.android.com/kotlin/style-guide#braces) it's not clear to me 🤷‍♂️ .
> Braces are otherwise required for any if, for, when branch, do, and while statements and expressions, even when the body is empty or contains only a single statement.
> ```kt
> if (string.isEmpty()) return  // WRONG
> else doLotsOfProcessingOn(string, otherParametersHere)
> ```
Ok, "WRONG". But why? The Kotlin style guide tells the reason 🙊 !🐑.

Anyways I've update this code since I noticed another improvement could be added by replacing the comment with a variable name that explains what goes on 👍 , which in turn allowed to write the condition as:
```kt
if (isInvalidQuery) R.string.a
else R.string.b
```

See commit: 3c3644e3

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainsViewModel.kt:
```diff
@@ -348,17 +336,18 @@ class SiteCreationDomainsViewModel @Inject constructor(
         )
     }
 
+    @Suppress("ForbiddenComment")
```
❓ Question: Is this needed here? I think there are no forbidden comments in this function.

ovitrif replied:
Good point 😉 , It's not needed anymore! I removed it in commit c96825a7

thomashorta commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/compose/DomainItem.kt:
```diff
@@ -0,0 +1,59 @@
+        Row(
+            modifier = Modifier
+                .clickable { onClick.invoke() }
+                .padding(horizontal = 24.dp, vertical = 16.dp)
```
❓ Question: I know we have a [Margin](https://github.com/wordpress-mobile/WordPress-Android/blob/7eca2b306f664b28c591de523974acbabfe0c920/WordPress/src/main/java/org/wordpress/android/ui/compose/unit/Margin.kt) class for some commonly used margin values, I wonder if it's worth using that or if it doesn't make sense for Compose. Wdyt?

Since we usually get the actual dimension value in Figma design specs instead of "tokens" (like: "use a medium margin" or "extra-large margin") I am not 100% sure what is the best approach.

ovitrif replied:
> ❓ Question: I know we have a [Margin](https://github.com/wordpress-mobile/WordPress-Android/blob/7eca2b306f664b28c591de523974acbabfe0c920/WordPress/src/main/java/org/wordpress/android/ui/compose/unit/Margin.kt) class for some commonly used margin values, I wonder if it's worth using that or if it doesn't make sense for Compose. Wdyt?

I agree with you, although there are other incoming changes which will most probably remove or change these margins (see task: #17985), and because of that my reasoning was that it's most simple to just use the actual values for now.

I would normally use the `Margin` class although my long term goal for it would be to turn it into an inline class but that's another topic for another time 👍 

> Since we usually get the actual dimension value in Figma design specs instead of "tokens" (like: "use a medium margin" or "extra-large margin") I am not 100% sure what is the best approach.

On a sidenote I keep thinking about this as well and dimensions aren't even the greatest issue I see when it comes to consistency in designs 😅 . Colors and typography are even more variable, for them almost every design in Figma redefines like 3+.

I would actually be in favor of just going with actual values and gradually stop creating so many "standards" in code for UI which ultimately change in the blink of an eye with every new design and end up being used only 2 times or less. Perhaps being flexible is the better approach here, and if that's to be our goal, then IMHO there's more flexibility in just typing the values directly.

As a general rule, the way I reason about this is that something has to be used at least 2 times to make it a variable or function for it, otherwise unless it helps make the code more readable, inlining is better than extracting IMHO.

thomashorta commented on WordPress/src/test/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainsViewModelTest.kt:
```diff
@@ -408,13 +408,18 @@ class SiteCreationDomainsViewModelTest : BaseUnitTest() {
             "invaliddomain"
         }
+
+    private fun mockDomain(name: String = "", free: Boolean = true) = mock<DomainModel> {
+        on { domainName } doReturn name
+        on { isFree } doReturn free
+    }
```
💡 Suggestion: Instead of creating a `mock`, could an actual instance of `DomainModel` be used here instead?

It might be a bit of nit-picking but I believe `mock` performance is a bit worse than using actual objects so I try to use class instances whenever possible.

ovitrif replied:
Actually I prefer mocks in unit tests, and performance is not an issue in our suite imho 😇. Building them takes time, but running is really fast 👍 

I will keep it a mock since it's unit tests code and the performance goals are different.

ovitrif committed:
Revert unwanted change

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainsViewModel.kt:
```diff
@@ -348,17 +336,18 @@ class SiteCreationDomainsViewModel @Inject constructor(
         )
     }
 
+    @Suppress("ForbiddenComment")
```
I think the `Suppress` annotation is not needed in this function

ovitrif replied:
Indeed, Thomas noticed it as well 👍 
Thanks, I removed it in commit c96825a7

thomashorta reviewed (commented):
Nice job! I added a few comments (mostly questions and minor suggestions).

Also, while testing it I noticed a couple of things and just wanted to confirm with you that this is the intender behavior right now.

1. When following this step: "Replace the `t` in the search field with `%`" both in the Control and Treatment Variations, I found that I only get the error if there are ONLY invalid characters, is that correct?
2. In the **Treatment Variation**, selecting an item doesn't make it "selected" (highlighted or something visually different), is that on purpose for now?


Screenshots for question 1:

| Valid + Invalid Chars | Only Invalid Chars |
| --- | --- | 
| ![image](https://user-images.githubusercontent.com/5091503/220728078-43390b0e-d5aa-405f-a7e8-06b2b69b7a06.png) | ![image](https://user-images.githubusercontent.com/5091503/220728174-02802112-6b3f-46b2-8c5b-b135252cdf29.png) |

thomashorta reviewed (changes requested):
🛑  Uh-oh, I was testing it a bit more and I found a crash 😬

First, I found that the animations get janky when tapping a free domain and that the keyboard is shown again even though I just hid it.

Then, I found out that when tapping 2 different free domains alternately **quickly**, the app crashes. 

It seems it happens because this changes the `selectedDomain` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/1f66e55bcae8fc2f503da121c5782bf2e9549feb/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainsViewModel.kt#L73-L77), causing an UI state update, which triggers [this observable here](https://github.com/wordpress-mobile/WordPress-Android/blob/a547fc19784032f7b45e9ee5e7216b901ce01b5a/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainsFragment.kt#L86-L95) and causes the whole UI to be rebuilt.

This causes the recycler view holder to call `onBind` again, but if done too quickly it seems to be causing a call to a `LifecycleOwner` that no longer exists (or something like that) when calling [this line](https://github.com/wordpress-mobile/WordPress-Android/blob/45eb7629d472e31b37a6f388db4f248090964fc1/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/domains/SiteCreationDomainViewHolder.kt#L74).

<details>
<summary>Crash video (OnePlus 5T on Android 10) ↘️</summary>

https://user-images.githubusercontent.com/5091503/220734985-21e6c213-28d2-44b8-8f6c-4bec93441da8.mp4
</details>

<details>
<summary>See the crash logs below ↘️</summary>

```js
FATAL EXCEPTION: main
Process: com.jetpack.android.prealpha, PID: 15414
java.lang.IllegalStateException: View tree for androidx.compose.ui.platform.ComposeView{9a4a4be V.E...... .......D 0,0-1080,142 #7f0a019b app:id/compose_view} has no ViewTreeLifecycleOwner
	at androidx.compose.ui.platform.ViewCompositionStrategy$DisposeOnViewTreeLifecycleDestroyed.installFor(ViewCompositionStrategy.android.kt:109)
	at androidx.compose.ui.platform.AbstractComposeView.setViewCompositionStrategy(ComposeView.android.kt:140)
	at org.wordpress.android.ui.sitecreation.domains.SiteCreationDomainViewHolder$DomainComposeItemViewHolder.onBind(SiteCreationDomainViewHolder.kt:74)
	at org.wordpress.android.ui.sitecreation.domains.SiteCreationDomainsAdapter.onBindViewHolder(SiteCreationDomainsAdapter.kt:40)
	at org.wordpress.android.ui.sitecreation.domains.SiteCreationDomainsAdapter.onBindViewHolder(SiteCreationDomainsAdapter.kt:19)
	at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7254)
	at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:7337)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:6194)
	at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6460)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6300)
	at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6296)
	at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2330)
	at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1631)
	at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1591)
	at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:668)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4309)
	at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:4012)
	at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4578)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1855)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at androidx.core.widget.NestedScrollView.onLayout(NestedScrollView.java:1983)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1103)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at com.google.android.material.appbar.HeaderScrollingViewBehavior.layoutChild(HeaderScrollingViewBehavior.java:148)
	at com.google.android.material.appbar.ViewOffsetBehavior.onLayoutChild(ViewOffsetBehavior.java:43)
	at com.google.android.material.appbar.AppBarLayout$ScrollingViewBehavior.onLayoutChild(AppBarLayout.java:2119)
	at androidx.coordinatorlayout.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:918)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1103)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at com.android.internal.policy.DecorView.onLayout(DecorView.java:786)
	at android.view.View.layout(View.java:22085)
	at android.view.ViewGroup.layout(ViewGroup.java:6290)
	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:3363)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2834)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1947)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8048)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1163)
	at android.view.Choreographer.doCallbacks(Choreographer.java:986)
	at android.view.Choreographer.doFrame(Choreographer.java:902)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1148)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7697)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
```
</details>

ovitrif committed:
Remove unnecessary forbidden comment suppression

ovitrif commented:
Thanks for your great review @thomashorta 🙇 

> 1. When following this step: "Replace the `t` in the search field with `%`" both in the Control and Treatment Variations, I found that I only get the error if there are ONLY invalid characters, is that correct?

Yep that's the expected and existing behavior on `trunk` or prod. I guess the backend does some sort of sanitization and it return matches for that sanitized query when it's not empty.

📮 **Postman** screenshots when querying the `/domains/suggestions/` GET endpoint:

<details>
<summary>Results for <code>query: t%</code></summary>

<img width="1195" alt="apiRes" src="https://user-images.githubusercontent.com/4588074/220863521-2ccca43d-8b53-4106-a56e-c5c55246b895.png">
</details>

<details>
<summary>Error for <code>query: %x</code></summary>

<img width="1195" alt="apiErr" src="https://user-images.githubusercontent.com/4588074/220864165-07b92163-2652-43cc-84fd-409a611f5aae.png">
</details>


> 2. In the **Treatment Variation**, selecting an item doesn't make it "selected" (highlighted or something visually different), is that on purpose for now?

True, that's because we were still deciding on the new UI, and this task was the initial plan: https://github.com/wordpress-mobile/WordPress-Android/issues/17901 where we would have shown a modal instead of any a visual indication on the selected item 😉 .
This PR is only concerned with the specs from the related task: https://github.com/wordpress-mobile/WordPress-Android/issues/17899 👍

ovitrif commented:
> 🛑 Uh-oh, I was testing it a bit more and I [found a crash](https://github.com/wordpress-mobile/WordPress-Android/pull/17998#pullrequestreview-1310009792) 😬

Thanks a lot for finding this bug @thomashorta before we merge the PR. It almost slipped through and we could have missed it 🤦 . This really reminds me how important it is to a little more exploratory testing in every PR and continue to dedicate the last week of a project to testing, a ritual we started last year in our team.

I was able to repro it and proceeded to address the issue in this commit: 039ea863. It doesn't reproduce for me anymore 🤞  

Thanks for investigating the call stack leading to it.
P.S. Please don't feel like you have to do this on my PRs. It's something I have to own and I'm totally ok if you just say: "_Hey Ovi, look, your PR might have introduced this bug that I can repro with these steps: [..]_".

ovitrif committed:
Refactor: Rename query update functions

ovitrif commented:
> I found that the animations get janky

@thomashorta Can you make a recording of this or explain how I could repro this?
I couldn't repro it myself.

Thanks in advance 🙏 


> the keyboard is shown again even though I just hid it

Another great discovery @thomashorta 🥇 . This seems to be a bug in production as well, I filed issue #18004 and added it on the project board.

I won't fix it it this PR since it's already an existing issue and this PR is starting to be concerned with too many things 😅 .

ovitrif committed:
Refactor: Rename view-holder classes to separate the old vs new

ovitrif committed:
Refactor: Put parameters on one line in view-holders

ovitrif committed:
Refactor: Rename ERROR_FETCH to ERROR_V1

ovitrif committed:
Refactor: Cleanup empty list message logic to get rid of the comment

ovitrif committed:
Refactor: Remove braces from if statement in `DomainsDiffUtils`

ovitrif committed:
Fix compose crash when selecting items quickly by moving config to init

ovitrif commented:
👋 @antonis (and @thomashorta) This PR is ready for another pass.

The [crash](https://github.com/wordpress-mobile/WordPress-Android/pull/17998#issuecomment-1441456956) has been resolved in 039ea86. I don't see any unaddressed issues except https://github.com/wordpress-mobile/WordPress-Android/issues/18004 which is in production as well so we skip it on purpose 🤞

thomashorta commented:
> @thomashorta Can you make a recording of this or explain how I could repro this? I couldn't repro it myself.
> 
> Thanks in advance 🙏

The original crash video I posted shows the animation issue, but it is subtle.

The animation I'm referring to is the ripple effect on item clicks when tapping a free domain that is not the currently selected one. It is very subtle so maybe not worth taking a look now, especially because I think it might be related to using Compose inside a `RecyclerView` and you added some TODOs regarding updating Compose and `RecyclerView` versions for better support.

Here's a video that makes the issue clearer by tapping twice on the same item. Notice how the first animation stops abruptly and the second one (which doesn't trigger a UI update) has a smoother fade-out. I also slowed down the animations using Developer Settings so it's easier to see, in the second part of the video.

https://user-images.githubusercontent.com/5091503/220904721-989ff6ab-77e8-428b-bb91-bc56b2a61235.mp4

But again, I'm fine with keeping it this way for now, as it is probably related to the `RecyclerView` + Compose interaction with the lib versions we're using.

thomashorta reviewed (approved):
LGTM and the crash doesn't happen anymore. Thanks!

antonis reviewed (approved):
Thank you for the update @ovitrif and for the extensive review @thomashorta 🙇 
> The https://github.com/wordpress-mobile/WordPress-Android/pull/17998#issuecomment-1441456956 has been resolved in https://github.com/wordpress-mobile/WordPress-Android/commit/039ea8636d813eac8d8376b883182e641f1592c2. I don't see any unaddressed issues except https://github.com/wordpress-mobile/WordPress-Android/issues/18004 which is in production as well so we skip it on purpose 🤞

I've retested with a debug build on my Pixel 5 / Android 13 and wasn't able to reproduce the crash or any other new issue. The code after the updates also LGTM 🎉

ovitrif commented:
> Here's a video that makes the issue clearer by tapping twice on the same item. Notice how the first animation stops abruptly and the second one (which doesn't trigger a UI update) has a smoother fade-out. I also slowed down the animations using Developer Settings so it's easier to see, in the second part of the video.

Thanks @thomashorta. It's rather inconsistent than "janky" IMHO.
Yeah, very possible it's related to using a compose view inside a recycler view 🤷 .
Doesn't feel like a blocker to me neither 👍

ovitrif commented:
Gonna merge this once [the PR of the base branch](https://github.com/wordpress-mobile/WordPress-Android/pull/17977) can be merged to `trunk`, thanks everyone for the feedback 🙇
